### ISSUE TITLE
feature/set up revenue tags for reporting

### ISSUE THREAD

**Issue by chris-dare** [serenity-health/roadmap#101]
**Tell us about your request!**
Providers need set the custom categories for their services and products. We can acheive this with revenue tags

**Which app does this concern?**
Provider portal

**What will be benefit of this feature?**
Revenue tags will help providers report on their bills the way they like

**Describe the solution you'd like**
On the API:
1. Create a model called `RevenueTag`. It should extend `ResourceModel` and have the following fields:
a. `display`: (unique and case insensitive) - Charfield with maxlength 256 (This will be the name of the revenue tag)
b. `created_by`: FK to `Authenticationmanager.User`
c. `is_core`: boolean with a default of `false` (each provider will have core default revenue tags which cannot be deleted. We will use this flag to identify which tags these are)
d. `Organization`: FK to `Organization`

2. Update the `Organization.create` method such that if the type of the organization is `PROV`, then the following revenue tags will be created:
a. Name="Other", created_by=self.owner, organization=self, is_core=True, organization=self  
b. Name="Pharmacy", created_by=self.owner, organization=self, is_core=True, organization=self  

3. Extend the `Healthcareservice`, `Stock` and `ChargeItem` model with the following properties:
a. revenue_tag: FK to `RevenueTag`  
b. revenue_tag_display: Charfield with maxlength 256  
Whenever instances of the Healthcareservice model are saved, do the following:  
a. if revenue_tag_display is not None, get `revenue_tag` with the display of `revenue_tag_display` and set it to `self.revenue_tag`     
b. else get the revenue tag of that organization with the display of `Other` and set that `revenue_tag` and `revenue_tag_display`  

3. Extend the HealthcareserviceSerializer to expose the `revenue_tag_display` property
Whenever instances of the Stock model is saved, do the following:  
a. if revenue_tag_display is not None, get `revenue_tag` with the display of `revenue_tag_display` and set it to `self.revenue_tag`    
b. else get the revenue tag of that organization with the display of `Pharmacy` and set that `revenue_tag` and `revenue_tag_display`  

Whenever instances of the ChargeItem model is created, do the following:
a. If the service category is medication, get the `revenue tag` of the related stock item and set the `revenue_tag` and `revenue_tag_display` of the charge item instance  
b. else get the revenue tag of the related Healthcare service and set the `revenue_tag` and `revenue_tag_display` of the charge item instance  
c. if these throw an error, then get the revenue tag of that organization named `Other` and set that `revenue_tag` and `revenue_tag_display`  
** Implement this behavior in a method named `set_revenue_tag` so can be called on the charge item instance

NB: You will need to extend the ChargeItem model to have a foreign key to the Stock model to simply this task. It will also help with other features and BI moving forward

Be sure to catch necessary exceptions

5. Implement a API view for provider revenue tags
a. This endpoint should be registered under `provider.valueset.views` and implemented like `diagnostics.views.DiagnosticTestCodeValuesetListView`
b. In addition to the above, you need to filter and return revenue tags that belong to a specific provider
c. Register this API view in `provider.valueset.urls` with the path name `/<provider_id>/revenue-tag-type`
d. Document this API on readme under the provider portal in a new section titled "valuesets"
The endpoint will be `{{base_url}}/providers/<provider_id>/revenue-tag-type`

NB: Be sure to remove the unique_together constraint in the healthcare service migration that will raise a bug. Ref: #82 

On the frontend,
Extend Healthcare service create and edit use cases such it it allows users to save the revenue tag based on values from `{{base_url}}/providers/<provider_id>/revenue-tag-type`

API endpoint can be found [here](https://serenity-health.readme.io/reference/valuesets)

**Describe alternatives you've considered**
N/A

**Additional context**
N/A

**Comment by chris-dare** [comment]
Hi @TuyizereBapt I'm yet to complete the description for this issue and will notify you once complete

**Comment by chris-dare** [comment]
Hi @TuyizereBapt,

I'm done updating the issue. This is a task due next Tuesday

**Comment by chris-dare** [comment]
@TuyizereBapt @seddie62 do you have any questions on this task?

**Comment by TuyizereBapt** [comment]
Hey @dexios1 what should happen for new HealthCareServices, Stocks and ChargeItems for already existing provider organizations without RevenueTags set ye?

**Comment by chris-dare** [comment]
> Hey @dexios1 what should happen for new HealthCareServices, Stocks and ChargeItems for already existing provider organizations without RevenueTags set ye? 

For `HealthcareService`, the revenue tage display default should be "Other"

For `Stock`, it should be "Pharmacy"

For ChargeItem, it should be None

**Comment by TuyizereBapt** [comment]
> > Hey @dexios1 what should happen for new HealthCareServices, Stocks and ChargeItems for already existing provider organizations without RevenueTags set ye?
> 
> For `HealthcareService`, the revenue tage display default should be "Other"
> 
> For `Stock`, it should be "Pharmacy"
> 
> For ChargeItem, it should be None

I get that. But remember it's just revenue_tag_display field only, there is also revenue_tag ForeignKey field. Do you mean this foreign key can be set to null in such a case?

**Comment by TuyizereBapt** [comment]
Done. Up for review now

Here is the [PR](https://github.com/serenity-health/core-systems-api/pull/45)

**Comment by chris-dare** [comment]
Noted. Will check it out

**Comment by seddie62** [comment]
@dexios1 have you merged it?

**Comment by chris-dare** [comment]
Just did


Kind regards,
Chris Dare

E: ***@***.******@***.***>
P: +233 26 226 0986 / +250 786 745 117<tel:+233%2026%20226%200986%20/%20+250%20786%20745%20117>
W: http://chrisdare.me<http://chrisdare.me/>
[LinkedIn]<https://linkedin.com/in/csdare/>[Twitter]<https://twitter.com/chrisdare_>


________________________________

**Comment by chris-dare** [comment]
Hi @seddie62 please let me know when this is integrated on the frontend

**Comment by seddie62** [comment]
@dexios1 @TuyizereBapt what is the api needed here to get all possible values and what is the field to be added

**Comment by chris-dare** [comment]
> > > Hey @dexios1 what should happen for new HealthCareServices, Stocks and ChargeItems for already existing provider organizations without RevenueTags set ye?
> > 
> > 
> > For `HealthcareService`, the revenue tage display default should be "Other"
> > For `Stock`, it should be "Pharmacy"
> > For ChargeItem, it should be None
> 
> I get that. But remember it's just revenue_tag_display field only, there is also revenue_tag ForeignKey field. Do you mean this foreign key can be set to null in such a case?

@TuyizereBapt sorry, I missed this. I think it depends on the final implementation. If the default revenue tags are set up, then there's no need for nullables, but if they aren't, then there's a need for that.

**Comment by chris-dare** [comment]
> @dexios1 @TuyizereBapt what is the api needed here to get all possible values and what is the field to be added

Hi @seddie62,

the field's key is `revenue_tag_display`. Please give this a label of "Revenue tag"

This is the endpoint to get all possible values: `{{base_url}}/providers/<provider_id>/revenue-tag-type`

@TuyizereBapt please share the link to the API spec with @seddie62

**Comment by TuyizereBapt** [comment]
> > @dexios1 @TuyizereBapt what is the api needed here to get all possible values and what is the field to be added
> 
> Hi @seddie62,
> 
> the field's key is `revenue_tag_display`. Please give this a label of "Revenue tag"
> 
> This is the endpoint to get all possible values: `{{base_url}}/providers/<provider_id>/revenue-tag-type`
> 
> @TuyizereBapt please share the link to the API spec with @seddie62

Here is the API spec: https://serenity-health.readme.io/reference/valuesets

**Comment by seddie62** [comment]
is the field required? @dexios1 @TuyizereBapt

**Comment by seddie62** [comment]
<img width="635" alt="Screenshot 2022-03-10 at 6 25 19 AM" src="https://user-images.githubusercontent.com/32453214/157602058-a8586d66-766c-4cfb-b1de-3ed8b4adbbe9.png">
<img width="632" alt="Screenshot 2022-03-10 at 6 25 11 AM" src="https://user-images.githubusercontent.com/32453214/157602075-3741a0a0-0918-4572-9007-d44167d20a72.png">

So I'm guessing this hasn't been merged? @dexios1

**Comment by TuyizereBapt** [comment]
Here is the API spec: https://serenity-health.readme.io/reference/valuesets

**Comment by seddie62** [comment]
@TuyizereBapt I'm referring to the actual field revenue_tag_display on the services API, is that required?

**Comment by TuyizereBapt** [comment]
No. It will be set to default value if you do not provide a `revenue_tag_display` value on creation

**Comment by chris-dare** [comment]
Hi @seddie62 @TuyizereBapt,

I'm just reviewing the API for the revenue tag valueset. @TuyizereBapt it needs to be `{{base_url}}/providers/{{serenity_user_provider_id}}/valueset/revenue-tags` and registered in `provider.valueset.urls`

**Comment by chris-dare** [comment]
> @TuyizereBapt I'm referring to the actual field revenue_tag_display on the services API, is that required?

Yes, Jean is right the API will do that. 
@seddie62 what you need to do on the frontend is to set the revenue tag to the default of "Other" for healthcare services. You need to do this so that users are not confused on how the defaults will be set up

@pkdadson for stock items, this will be "Pharmacy"

**Comment by TuyizereBapt** [comment]
> 

I don't that url is possible considering that the sub paths for valueset under provider do not have <provider_id>

Example: provider.urls - `path("/valueset/", include("provider.valueset.urls")),`

For the above url to work, it would require the above url to `path("/<provider_id>/valueset/", include("provider.valueset.urls")),`

**Comment by TuyizereBapt** [comment]
> I don't think that url is possible considering that the sub paths for valueset under provider do not have <provider_id>
> 
> Example: provider.urls - `path("/valueset/", include("provider.valueset.urls")),`
> 
> For the above url to work, it would require the above url to `path("/<provider_id>/valueset/", include("provider.valueset.urls")),`

I don't think that url is possible considering that the sub paths for valueset under provider do not have <provider_id>

Example: provider.urls - `path("/valueset/", include("provider.valueset.urls")),`

For the above url to work, it would require the above url to be altered to `path("/<provider_id>/valueset/", include("provider.valueset.urls")),`

**Comment by chris-dare** [comment]
> > I don't think that url is possible considering that the sub paths for valueset under provider do not have <provider_id>
> > Example: provider.urls - `path("/valueset/", include("provider.valueset.urls")),`
> > For the above url to work, it would require the above url to `path("/<provider_id>/valueset/", include("provider.valueset.urls")),`
> 
> I don't think that url is possible considering that the sub paths for valueset under provider do not have <provider_id>
> 
> Example: provider.urls - `path("/valueset/", include("provider.valueset.urls")),`
> 
> For the above url to work, it would require the above url to be altered to `path("/<provider_id>/valueset/", include("provider.valueset.urls")),`

Ah...got it. Alright. Then this is fine. Thanks a lot!
